\documentclass[final]{siamart171218}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{bm}

\setlength{\oddsidemargin}{0.65in}
\setlength{\evensidemargin}{0.65in}

\title{The ``term-count log-ratio'' statistic for topic modeling analysis
  of differential gene expression}

\author{Peter Carbonetto\thanks{Dept. of Human Genetics and the Research Computing Center, University of Chicago, Chicago, IL}}

\begin{document}

\maketitle

\section{Differential gene expression}

The ``log-fold change'' statistic is commonly used in microarray and
RNA sequencing experiments to quantify expression changes between two
conditions (e.g., \cite{cui-churchill-2003, quackenbush-2002}). To
motivate the ideas below, I write the log-fold change for gene $j$
and condition $k$ as a ratio of two conditional expectations,
\begin{equation}
\mathsf{lfc}(j,k) = \log_2 \frac{E[x_j \,|\, \mathrm{condition} = k]}
                                {E[x_j \,|\, \mathrm{condition} \neq k]},
\end{equation}
where $x_j$ is the measured expression level (e.g., UMI count) of gene
$j$. In experiments where the conditions are inferred---for example,
by running a machine learning algorithm to cluster the expression
profiles---this quantity could represent the difference in gene
expression between cells inside and outside a cluster.

Supposing $n_k$ out of a total of $n$ gene expression profiles
(cells) are from condition $k$, then $\mathsf{lfc}(j,k)$ can be
computed as
\begin{equation}
\mathsf{lfc}(j,k) =
\log_2 \left\{\frac{n_{jk}}{n_j - n_{jk}} \times \frac{n - n_k}{n_k} \right\},
\end{equation}
where $n_j$ is the total expression of gene $j$ among all expression
profiles, and $n_{jk}$ is the total expression of $j$ among all cells
in condition (or cluster) $k$.

The aim in the next sections is to define a analogue to the log-fold
change statistic for topic modeling.

\section{Poisson non-negative matrix factorization and the multinomial
  topic model}

Here we briefly describe the multinomial topic model and Poisson
non-negative matrix factorization---these are covered in much more
detail in other documents.

The topic model describes a process for generating an $n \times m$
matrix of counts, $X$. We begin with the ``bag of words'' description,
which is what is used to describe LDA \cite{blei-2003}. In this view,
each row $i$ is a document (or gene expression profile), and let $m_i$
be the size of this document; that is, $m_i = \sum_{j=1}^m
x_{ij}$. The vector $w_i$ is a vector of terms (or genes) of length
$m_i$ (the order of the words or genes appearing in this vector
doesn't matter, hence the ``bag of words''). For each $t = 1, \ldots,
m_{i}$, the word/gene $w_{it}$ is equal to $j$ with probability
$p(w_{it} \,|\, z_{it} = k) = f_{jk}$, where $z_{it}$ is a variable
indicating which topic, $k \in \{1, \ldots, K\}$ the word/gene belongs
to. The topic indicator variable is in turn generated according to
$p(z_{it} = k) = l_{ik}$, where $l_{i1}, \ldots, l_{iK}$ is a
document-specific probability table.

Returning to the question of assessing differential gene expression,
there are two ``twists'' relative to the standard analysis: one, the
group (topic) assignments are probabilistic; two, the group
assignments are made at the level of genes, not cells. With these two
points in mind, I propose the ``term-count log-ratio'',
\begin{equation}
\mathsf{tclr}(j,k) = \log_2 \left\{ \frac{x}{y} \right\}.
\end{equation}

\bibliographystyle{siamplain}
\bibliography{tclr}

\end{document}

