---
title: "Illustration of fastTopics applied to a single-cell RNA-seq data set"
author: "Peter Carbonetto"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{fastTopics-intro}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE,results = "hold",comment = "#",
                      fig.align = "center",fig.height = 4,fig.width = 6)
```

In this brief vignette, we illustrate the use of the alternating
sequential quadratic programming ("alt-SQP") method, implemented in the
`fastTopics` package, for fitting a non-negative matrix factorization
to read count data from a single-cell RNA-seq experiment. We compare
the runtime and accuracy of alt-SQP against the commonly used
multiplicative update rules, which can be viewed as an EM algorithm.

## Set up environment

We begin with the initial setup steps. First, load the packages used in
the example below.

```{r load-pkgs, message=FALSE}
library(Matrix)
library(fastTopics)
```

Next, initialize the pseudorandom number generator.

```{r set-seed}
set.seed(1)
```

## Load the data set

Load the "droplet" read count data---these are gene expression
profiles of trachea epithelial cells in C57BL/6 mice obtained using
droplet-based 3' single-cell RNA-seq; see [Montoro *et al*
(2018)][montoro-paper] for details. The counts matrix contains gene
expression data (read counts) for 17,133 genes collected from 7,193
samples. The majority of the counts are zero.

```{r load-data}
data(droplet)
n <- nrow(droplet)
m <- ncol(droplet)
n
m
mean(droplet > 0)
```

## Fit non-negative matrix factorization

This is the number of factors ("topics") in the non-negative matrix
factorization. Montoro *et al* assigned 13 unique labels ("cell
types"), so this is the number of topics factors we will use in our
analysis.

```{r init-k}
k <- 13
```

Create a random initialization of the factors and loadings.

```{r init-factors-and-loadings}
F <- matrix(runif(m*k),m,k)
L <- matrix(runif(n*k),n,k)
```

Let's now try running 500 multiplicative (EM) updates. These updates
are very simple and fast, so we can run a lot of theem in a relatively
short amount of time. The EM algorithm is recovered by turning off the
SQP updates in `altsqp`. (You may want to adjust the `nc` control
parameter to match the number of available cores on your computer.)

```{r run-betanmf}
fit1 <- altsqp(droplet,list(L = L,F = F),numiter = 500,
               control = list(numsqp = 0,extrapolate = Inf,nc = 4))
```

Now try running 20 SQP updates (with extrapolation starting at
iteration 20). Each SQP update is much slower than EM, but, as we will
see, each SQP iteration makes much more progress toward a solution.

```{r run-altsqp}
fit2 <- altsqp(droplet,list(L = L,F = F),numiter = 40,
               control = list(extrapolate = 20,nc = 4))
```

Indeed, the alt-SQP solution is much better (higher likelihood) than
the EM solution:

```{r evaluate-fits}
print(loglik.poisson(droplet,fit1),digits = 16)
print(loglik.poisson(droplet,fit2),digits = 16)
```

And it took less effort for alt-SQP to arrive at the better solution
(these two numbers are the runtimes in seconds):

```{r compare-runtimes}
sum(fit1$progress$timing)
sum(fit2$progress$timing)
```

In practice, you will want to run more SQP updates to get a better
solution---this is only an illustration. In practice, somewhere
between 100 and 200 iterations is good enough for solutions that
nearly satisfy the KKT optimality conditions.

## Get topic memberships

The output of `altsqp` is a non-negative matrix factorization. By
calling `poisson2multinom`, we can recover a grade-of-membership
model, in which the outputted `L` and `F` are maximum-likelihood
estimates of the "loadings" (topic probabilities) and "factors" (gene
probabilities).

```{r get-multinom-fit}
fit1 <- poisson2multinom(fit1)
fit2 <- poisson2multinom(fit2)
```

If we measure the quality of the grade-of-membership (multinomial)
model fits, the alt-SQP estimates again come out on top:

```{r evaluate-multinom-fit}
print(loglik.multinom(droplet,fit1),digits = 16)
print(loglik.multinom(droplet,fit2),digits = 16)
```

## Session information

This code chunk gives information about the computing environment used
to generate the results contained in this vignette, including the
version of R and the packages used.

```{r session-info}
sessionInfo()
```

[montoro-paper]: http://dx.doi.org/10.1038/s41586-018-0393-7
