---
title: "Analysis of single-cell RNA-seq data using fastTopics: an illustration"
author: Peter Carbonetto
date: "`r Sys.Date()`"
output:
  html_document:
    theme: readable
vignette: >
  %\VignetteIndexEntry{Analysis of single-cell RNA-seq data using fastTopics: an illustration}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Here we illustrate the use of **fastTopics** to infer structure from
single-cell RNA-seq data. A basic analysis consists of the following
steps:

1. Load and prepare the count data.

2. Fit a Poisson non-negative matrix factorization (NMF) to the UMI
   count data.

3. Recover the multinomial topic model---also called the multinomial
   "grade-of-membership" model---from the Poisson NMF model fit.

4. Analyze structure in the estimated mixture proportions matrix.

5. Annotate selected clusters and factors by examining genes that
   are highly differentially expressed among clusters or factors.

In this vignette, we walk through each of these steps, and demonstrate
how fastTopics is used to implement the steps.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Begin by loading the packages used below. The ggplot2 and cowplot
packages are used to create some of the plots.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Matrix)
library(fastTopics)
library(ggplot2)
library(cowplot)
```

Set the seed so that the results can be reproduced.

```{r set-seed}
set.seed(1)
```

Load and prepare data
---------------------

Load the single-cell RNA-seq data:

```{r load-data}
data(pbmc_4k)
samples <- pbmc_4k$samples
genes   <- pbmc_4k$genes
counts  <- pbmc_4k$counts
dim(counts)
```

This data set is made up of reference transcriptome profiles that were
generated from 10 different bead-enriched subpopulations of PBMCs
(Donor A), as described in [Zheng *et al* (2017)][zheng-2017]. The UMI
counts are stored as an $n \times m$ sparse matrix, where $n$ is the
number of cells and $m$ is the number of genes.

No normalization or transformation of the UMI counts is needed. The
only pre-processing step we take is to remove genes that are not
expressed in any of the cells.

```{r filter-genes}
j      <- which(colSums(counts > 0) >= 1)
genes  <- genes[j,]
counts <- counts[,j]
dim(counts)
```

As expected, most (>95%) of the UMI counts are zero:

```{r nonzeros}
mean(counts > 0)
```

Fit Poisson NMF model
---------------------

By exploiting the relationship between the multinomial topic model and
non-negative matrix factorization (we have a vignette elaborating on
this relationship), we leverage fast NMF optimization algorithms to
fit the multinomial topic model.

We chose $K = 6$ factors. In practice, you will likely want to explore
the results from different settings of $K$.

Since the model fitting can take several minutes to run, or more, we
have saved a pre-fitted model:

```{r load-fit}
fit <- pbmc_4k$fit
```

If you would like to see the model fitting in action, run this code:

```{r fit-poisson-nm, eval=FALSE}
fit0 <- fit_poisson_nmf(counts,k = 6,numiter = 100,method = "em",
                        control = list(numiter = 4,nc = 2))
fit  <- fit_poisson_nmf(counts,fit0 = fit0,numiter = 100,method = "scd",
                        control = list(extrapolate = TRUE,numiter = 4,nc = 2))
```

(This is the code that was used to generate the stored result.) Before
running this code, may want to change `control$nc`. This controls the
number of threads in the TBB multithreaded computations, and for best
performance this should be set to the number of availablle processors.

Judging by this plot showing the change in the log-likelihood over
time, the iterates get close to a maxium-likelihood solution after
about 150 iterations.

```{r plot-loglik, fig.height=2, fig.width=3.5}
plot_progress_poisson_nmf(fit,x = "iter",add.point.every = 10,
                          colors = "black") +
  theme_cowplot(font_size = 10)
```

**Some general guidance on model fitting.** We have found that the EM
algorithm initially makes rapid progress toward a solution, but this
convergence slows considerably as the iterates approach the solution.
Close to a solution, we have found that other algorithms make much
more rapid progress than EM. Therefore, we have taken a two-stage
approach to model fitting: first, we perform EM updates; second, we
refine the fit using extrapolated SCD updates (we also tried
extrapolated CCD updates, and updates without extrapolation, and we
found that the extrapolated SCD updates almost always worked best).
For larger data sets, more than 200 updates may be needed to obtain a
good fit.

Recover multinomial topic model
-------------------------------

Once we have fit a Poisson NMF model, the multinomial topic model is
easly recovered by a reparameterization:

```{r poisson2multinom}
fit_multinom <- poisson2multinom(fit)
```

Identify clusters from mixture proportions
------------------------------------------

The key strength of fastTopics is its flexibility to capture a variety
of structure, from distinct clusters (e.g., cell types) to continuous
structure. Here, we show how the multinomial topic model fit can be
used to identify clusters of cells, and later we will expand on this.

Here we apply the most basic and most widely used clustering method,
*k*-means, to the mixture proportions matrix, $L$. We experimented
with different numbers of clusters and we found that 6 clusters worked
best in this example. Note that *k*-means is sensitive to (random)
initialization, so we set the seed.

```{r clustering-1}
set.seed(1)
clusters <- kmeans(fit_multinom$L,centers = 6,iter.max = 100)$cluster
```

In some cases, it may be better to apply *k*-means to the principal
components (PCs) of the mixture proportions. The PCs can easily be
computed using `prcomp`.

You will find that some visual inspection followed by manual
refinement may be needed to get the "right" clustering, regardless of
whether *k*-means or another clustering algorithm is used. Here, we
found that *k*-means split the largest cluster into two subclusters.
After merging these two subclusters, we ended up with 5 clusters:

```{r clustering-2}
clusters[clusters == 3] <- 1
clusters <- factor(clusters)
summary(clusters)
```

With the benefit of hindsight, we know that these 5 clusters
correspond to 5 cell types: B cells, T cells, natural killer (NK)
cellls, CD14+ monoytes and CD34+ cells. For clarity of presentation,
we label the clusters by these cell types, but noting that labeling
the clusters is not usually not possible until downstream analyses
have been conducted.

The clustering of the cells is mostly evident from the first two PCs of
the mixture proportions:

```{r pca-1, fig.width=6.25, fig.height=2, message=FALSE}
levels(clusters) <- c("T","CD14+","B","CD34+","NK")
cluster_colors <- c("gold","forestgreen","dodgerblue","darkmagenta","skyblue")
p1 <- pca_plot(fit_multinom,fill = clusters) +
  scale_fill_manual(values = cluster_colors)
p2 <- pca_hexbin_plot(fit_multinom)
plot_grid(p1,p2,rel_widths = c(9,10))
```

These PC plots help to illuminate the clustering and other
substructure.  The PC density plot is particularly helpful when the
plot contains thousands of cells, many of which overlap each other in
the scatterplog. Here we use a "hexbin" plot to visualize the density
of the cells projected onto the top two PCs.

From the density plot, observe that the NK cells and T cells form two
clusters that are less distinct than the others. In particular, there
is some ambiguity in how the cells should be assigned to the two
clusters.

The clustering of the B cells and CD34+ cells is more clear in PCs
3 and 4.

Summarize substructure with a Structure plot
--------------------------------------------

A much more versatile tool for exploring the mixture proportions is
the Structure plot. Here is the simplest example of a Structure plot:

```{r structure-plot-1, fig.width=6.5, fig.height=1.5, results="hide"}
topic_colors <- c("skyblue","forestgreen","darkmagenta","dodgerblue",
                  "gold","darkorange")
p3 <- structure_plot(fit_multinom,topics = 1:6,colors = topic_colors)
print(p3)
```

This is effectively a stacked bar chart, in which each bar corresponds
to a cell, and each colour corresponds to a factor, or "topic". Since
the mixture proportions for each cell sum to 1, the stacked bars all
have the same height of 1.

By default, `structure_plot` shows a random subset of 2,000 cells. The
cells are automatically ordered by computing a 1-d *t*-SNE embedding
on this random subset, then ordering the cells according to their
arrangement on this 1-d embedding. (There are options in
`structure_plot` to override this automatic behaviour.)

The colours were chosen to correspond to the clusters shown in the
plots above.

Here we see that a lot of information about the structure in the data
is compressed into a single plot. In particular, the B cells (blue),
CD14+ monocytes (green) and CD34+ cells (purple) are each largely
characterised by a single pattern of gene expression (a factor),
whereas expression in NK and T cells is less distinct, and is
reflected in the sharing of factor 6 among NK and T cells.

The versality of the Structure plot can also present a challenge,
particularly for cell populations with more complex or fine-scale
structure. Therefore, it may take some effort---and creativity---to
create a Structure plot that effectively highlights the structure in
the data. A simple way to improve the Structure plot is to arrange the
cells by their assignment to the 5 clusters:

```{r structure-plot-2, fig.width=6.5, fig.height=1.5, results="hide", message=FALSE}
p4 <- structure_plot(fit_multinom,topics = 1:6,colors = topic_colors,
                     grouping = clusters,gap = 25)
print(p4)
```

This second Structure plot helps to highlight heterogeneity within the
clusters, particularly within the T cells. Notably, within the T cell
cluster there is a subset of cells with a distinctive combination of
topics, in particular with sizeable contributions from the "NK cells"
factor (factor 1, light blue)---this is the same subset of cells that
sit in between the concentrated clumps of NK and T-cell in the PCA
hexbin plot above. To highlight this subset of cells, which turn out
to correspond to CD8+ cells, so we label this T cell subset as "CD8+".

```{r clustering-3}
clusters <- as.character(clusters)
clusters[clusters == "T" & fit_multinom$L[,1] > 0.1] <- "CD8+"
clusters <- factor(clusters)
summary(clusters)
```

This is the new Structure plot with the cells subdivided into 6
subsets:

```{r structure-plot-3, fig.width=6.5, fig.height=1.5, results="hide", message=FALSE}
p5 <- structure_plot(fit_multinom,topics = 1:6,colors = topic_colors,
                     grouping = clusters,gap = 25)
print(p5)
```

It is clear from this Structure plot that some subsets form distinct
clusters (CD14+, CD34+ and B cells) because expression factors
uniquely occur in these clusters, whereas others (CD8+, NK and T
cells) are less distinct, with somewhat arbitrary cluster boundaries,
and this is reflected in contributions of shared factors. For the less
distinct clusters, the clustering is nonetheless useful for guiding
the eye in the Structure plots toward interesting patterns.

Annotate factors and clusters using differential gene expression
---------------------------------------------------------------

Here is where we use `diff_count_analysis` to interpret selected
clusters and topics.

```{r volcano-plots, eval=FALSE}
ans <- diff_count_analysis(fit_multinom,counts)
volcano_plot(ans,k = 5,labels = genes$symbol)
```

TO DO: Demonstrate how to create interactive volcano plot.

[zheng-2017]: https://doi.org/10.1038/ncomms14049
