---
title: "Analysis of single-cell RNA-seq data using a topic model, Part 1: basic concepts"
author: Peter Carbonetto
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: no
    highlight: textmate
    theme: readable
vignette: >
  %\VignetteIndexEntry{Analysis of single-cell RNA-seq data using a topic model, Part 1: basic concepts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The aim of this vignette is to introduce the basic concepts behind
an analysis of single-cell RNA-seq data using a topic model. 

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

We begin our analysis by loading the Matrix and fastTopics
packages. Then we set the seed so that the results can be reproduced.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Matrix)
library(fastTopics)
set.seed(1)
```

The data set
------------

We will illustrate the concepts using a single-cell RNA-seq data set
from [Zheng *et al* (2017)][zheng-2017]. These data are reference
transcriptome profiles from 10 bead-enriched subpopulations of
peripheral blood mononuclear cells (PBMCs). The original data set is
much larger---for this introduction, we have taken a small subset of
approximately 3,700 cells.

The data we will analyze are the unique molecular identifier (UMI)
counts. These data are stored as an $n \times m$ sparse matrix, where
$n$ is the number of cells and $m$ is the number of genes:

```{r load-data}
data(pbmc_4k)
counts <- pbmc_4k$counts
dim(counts)
```

The UMI count data are expected to be "sparse"---that is, most of the
counts are expected to be zero. Indeed, over 95% of the UMI counts are
zero:

```{r nonzeros}
mean(counts > 0)
```

No need to normalize or transform
---------------------------------

Most analyses single-cell RNA-seq data involve a pre-processing step
in which the UMI counts are log-transformed and normalized. *When
analyzing UMI counts with a topic model, you should not normalize or
transform the counts.* In particular:

1. There is no need to log-transform the UMI counts because the counts
   are modeled directly using the multinomial distribution.

2. There is no need to normalize the counts in each cell by sequencing
   coverage because the multinomial is *conditioned* on the cell's
   total UMI count.

Additionally, any analyses select only the most highly variable
genes. *We recommend instead using all genes.* The only exception are
genes in which all the UMI counts are zero---these genes should be
removed prior to a topic model analysis.

Fit the topic model
-------------------

Since no pre-processing is needed, we can move directly to the next
step of the analysis: fitting the topic model to the UMI count
data. This is accomplished by a single call to `fit_topic_model`:

```{r fit-topic-model, eval=FALSE}
fit <- fit_topic_model(counts,k = 6)
```

This is typically the most computationally burdensome step---it may
take several minutes to run `fit_topic_model` on this data set. For
convenience, we saved the output from this call:

```{r load-fit}
fit <- pbmc_4k$fit
```

To fit a topic model, we must specify `k`, the number of topics.
Here, we have chosen `k = 6` topics. In most settings, a good choice
of `k` will not be known in advance, so you will you want to explore
the results from topic models at different settings of `k`.

There are several complexities to topic model fitting---we do not
elaborate on these complexities here. The `fit_topic_model` interface
is intended to hide most of these complexities, and it should work
well for a wide range of data sets. However, larger or more
challenging data sets my require some fine-tuning of the model
fitting. Model fitting is covered in more detail in a separate
vignette.

Each cell is represented as a unique mixture of the topics
----------------------------------------------------------

A key feature of the topic model is that each cell $i$ is represented
as a *unique mixture* of the topics. This cell-specific mixture is
learned from the data. Therefore, each cell can be summarized from the
$K$ mixture proportions. In fastTopics, the mixture proportions for
all cells are stored as an $n \times K$ matrix:

```{r loadings-1}
dim(fit$L)
```

To illustrate, here is a cell in which the pattern of expression is
almost fully captured by the fourth topic:

```{r loadings-2}
rows <- "GATATATGTCAGTG-1-b_cells"
round(fit$L[rows,],digits = 3)
```

Here are two more examples:

```{r loadings-3}
rows <- c("GACAGTACCTGTGA-1-memory_t",
          "TGAAGCACACAGCT-1-b_cells")
round(fit$L[rows,],digits = 3)
```

The third example is interesting because the observed expression is
best captured by a *mixture* of topics 4 and 6.

To make sense of these results, we first need to understand the
biological relevance of the topics. We turn to this question next.

Interpreting the topics using available cell labels
---------------------------------------------------

In some cases, you may have additional information about the cells,
such as the tissue the cells were sampled from. In the PBMC data set,
the cells were labeled using fluorescence-activated cell sorting
(FACS). There are five labels (cell types): B cells, CD14+ monocytes,
CD34+ cells, natural killer (NK) cells, and T helper cells.

```{r summary-subpop}
summary(pbmc_4k$samples$subpop)
```

We create a Structure plot to visualize the relationship between the
cell labels and the mixture proportions:

```{r structure-plot-with-celltype-labels, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
topic_colors <- c("skyblue","forestgreen","darkmagenta","dodgerblue",
                  "gold","darkorange")
p1 <- structure_plot(fit,topics = 1:6,colors = topic_colors,
                     grouping = pbmc_4k$samples$subpop,gap = 25)
print(p1)
```

The Structure plot is effectively a stacked bar chart, in which each
topic corresponds to a bar of a different colour. Being *proportions*,
the mixture proportions for each cell must sum to 1, so the total
height of the bars is the same for all cells, which makes it easier to
compare across cells. Patterns emerge when the cells are arranged so
that cells with similar mixture proportions are nearby each other
along the horizontal axis. The arranging of the cells within each cell
type is automated in `structure_plot` using the *t*-SNE method.

From the Structure plot, it is evident that topics 1, 2 and 4
correspond to the NK, CD14+ and B cell types; the expression in these
cells is largely explained by a single topic, although NK cells do
show more heterogeneity in expression, in which some cells "look like"
NK cells more than others.

In the CD34+ cells, we see that the mixture proportions contribute a
lot of useful information about these cells that wasn't captured by
the cell type labels alone. While expression in these cells is largely
explained by topic 3, some cells appear to be mislabeled---according
to the topic model, they should have been labeled as B cells or CD14+
cells, and there are other cell transcriptome profiles that are best
fit by a mixture of several topics.

Finally, the T cells subpopulation is another example where the topic
model captures interesting substructure within the cell type. In the
Structure plot, two patterns stand out: one, most T cells are a
mixture of two topics, 5 and 6, with a wide variation in the mixing
proportions for these two topics; there is a distinctive subset of T
cells that is captured by a mixture of three topics (topics 1, 5 and
6). As we will see shortly, we can analyze differences in gene
expression among the topics to help us understand what aspects of T
cell diversity are captured by these three topics.

Topics capture patterns of relative expression
----------------------------------------------

Above, we showed that cells are represented by mixtures of topics. In
our running example, some of the topics seem to be meaningful as they
correspond closely to the known cell types. However, we have not yet
explained what the topics *are* when analyzing single-cell RNA-seq
data. Each topic in fact captures a pattern of gene expression,
represented as a vector of $m$ relative expression levels ($m$ is the
number of genes). The relative expression levels for all topics are
stored as an $m \times K$ matrix:

```{r factors-1}
dim(fit$F)
```

Because these are estimates of *relative* expression, we can directly
compare these estimates across topics. For example, *CD79B* is
important to B cell differentiation, so the higher relative expression
level in topic 4 is a clue that this topic is picking up B cell
expression patterns (we multiplied the estimates by $10^6$ to make
them easier to read):

```{r factors-2}
j <- which(pbmc_4k$genes$symbol == "CD79B")
round(1e6 * fit$F[j,])
```

A much stronger clue is that the B cells marker gene *CD79A* is
uniquely expressed in topic 4:

```{r factors-3}
j <- which(pbmc_4k$genes$symbol == "CD79A")
round(1e6 * fit$F[j,])
```

By contrast, genes that are highly expressed in all topics, such as
gene *COX4I1*, provide little useful information:

```{r factors-4}
j <- which(pbmc_4k$genes$symbol == "COX4I1")
round(1e6 * fit$F[j,])
```

This points to the idea that we should not look for genes with high
relative expression levels, but rather genes that have much higher
relative expression levels than the other topics---and more generally,
genes with large expression differences. This idea---differential
expression analysis---is an old idea. The next section explains how
differential expression analysis is performed in the topic model.

Interpreting the topics by analyzing differences in gene expression
-------------------------------------------------------------------

To set the stage for differential expression analysis using a topic
model, we begin with a "classic" differential expression analysis
using the provided cell type labels. This analysis centers on
calculation of the "log-fold change" statistic, commonly denoted as
$beta$. It is the (base-2) log-ratio of expression (total UMI count)
among cells within and outside a group. Returning to the $CD79A$ and
$CD79B$ gene examples, we see that they are indeed highly
differentially expressed in B cells, roughly by a factor of aat least
$2^5 = 32$ times more expressed in cells labeled as "B cell" compared
to cells not labeled as such, and both genes are underexpressed in the
other cell types. So our analysis supports the notion that these genes
are very much characteristic of B cells.

```{r diff-count-analysis-1}
out1 <- diff_count_clusters(counts,pbmc_4k$samples$subpop)
rbind(out1$beta[pbmc_4k$genes$symbol == "CD79A",],
      out1$beta[pbmc_4k$genes$symbol == "CD79B",])
```

We have extended this differential expression analysis in the setting
where memberships to groups are *partial*, and this analysis is
implemented by the `diff_count_analysis` function:

```{r diff-count-analysis-2}
out2 <- diff_count_analysis(fit,counts)
```

As in the classic analysis, we get log-fold change estimates:

```{r diff-count-analysis-3}
rbind(out2$beta[pbmc_4k$genes$symbol == "CD79A",],
      out2$beta[pbmc_4k$genes$symbol == "CD79B",])
```

Both log-fold change estimates are noticeably larger in the "B cells"
topic (topic 4). In particular, the log-fold change estimate of 10 for
*CD79A* means that this gene is 1,000 times more expressed in topic 4
compared to other topics. This is probably an underestimate, as this
reflects that the gene is essentially not expressed outside the B
cells topics, but due to finite numeric calculations we run into
difficulties when computing ratios with zero denominators; so a
log-fold change statistic of 10 basically means "very large".

This is all well and fine for known genes.

Now we need some way of ranking genes in cases where we either want to
discover new genes for known cell types, or we do not yet know the
biological relevance of topic Intuitively, we might want to rank by
the log-fold change. The complication is that...

Volcano plot for the B cells topic:

```{r volcano-plot-bcells, fig.height=3.5, fig.width=4.5}
p2 <- volcano_plot(out2,k = 4,label_above_quantile = 0.998,
                   labels = pbmc_4k$genes$symbol,max.overlaps = 15)
print(p2)
```

Evaluating the topic model
--------------------------

```{r}
```

Old stuff
=========

A basic analysis consists of the following steps:

1. Load and prepare the count data.

2. Fit a Poisson non-negative matrix factorization (NMF) to the UMI
   count data.

3. Recover the multinomial topic model---also called the multinomial
   "grade-of-membership" model---from the Poisson NMF model fit.

4. Analyze structure in the estimated mixture proportions matrix.

5. Annotate selected clusters and factors by examining genes that
   are highly differentially expressed among clusters or factors.

In this vignette, we walk through each of these steps and demonstrate
how to implement these steps using fastTopics. In other exploratory
analyses, these steps might not be strictly sequential---for example,
the annotation in Step 5 may inform the exploration of substructure in
Step 4.

Identify clusters
-----------------

The topic or grade-of-membership model is able to capture a variety of
structure, from discrete (clusters) to continuous. We begin by showing
how the model can be used to identify clusters.

Here we apply the simplest clustering method, *k*-means, to the
mixture proportions matrix, $L$. We experimented with different
numbers of clusters and we found that 6 clusters worked best in this
example. Note that *k*-means is sensitive to (random) initialization,
so we set the seed.

```{r clustering-1, eval=FALSE}
set.seed(1)
clusters <- kmeans(fit_multinom$L,centers = 6,iter.max = 100)$cluster
```

Note that, in some cases, it may be better to apply *k*-means to the
principal components (PCs) of the mixture proportions, rather than the
mixture proportions directly. The PCs can easily be computed using
`prcomp`.

You will find that some visual inspection followed by manual
refinement may be needed to get the "right" clustering, regardless of
whether *k*-means or another clustering algorithm is used. By visual
inspection, we found that *k*-means split the largest cluster into two
subclusters. After merging these two subclusters, we ended up with 5
clusters:

```{r clustering-2, eval=FALSE}
clusters[clusters == 3] <- 1
clusters <- factor(clusters)
summary(clusters)
```

With the benefit of hindsight, we know that these 5 clusters
correspond to 5 cell types: B cells, T cells, natural killer (NK)
cellls, CD14+ monoytes and CD34+ cells. (Below we show how to annotate
these clusters by performing a differential expression analysis.) For
clarity of presentation, we label the clusters by these cell types,
noting that, in most practical settings, labeling the clusters is not
usually not possible until downstream analyses have been conducted.

The clustering of the cells is mostly evident from the top two PCs of
the mixture proportions:

```{r pca-1, fig.width=6.25, fig.height=2, message=FALSE, eval=FALSE}
levels(clusters) <- c("T","CD14+","B","CD34+","NK")
cluster_colors <- c("gold","forestgreen","dodgerblue","darkmagenta","skyblue")
p1 <- pca_plot(fit_multinom,fill = clusters) +
  scale_fill_manual(values = cluster_colors)
p2 <- pca_hexbin_plot(fit_multinom,bins = 24)
plot_grid(p1,p2,rel_widths = c(9,10))
```

Examining the projection of the cells onto PCs 1 and 2, the B cells
cluster is the most distinct, while the NK and T cell clusters are
much less distinct, and arguably it is incorrect to call them
"clusters". (The CD14+ and CD34+ cells separate much more clearly in
PCs 3 and 4.)

The PCA "hexbin" density plot is most helpful when there are many
overlapping data points in the PCA plot.

Visualize mixture proportions using a Structure plot
----------------------------------------------------

A much more versatile tool for exploring the mixture proportions is
the Structure plot. Here is a basic example:

```{r structure-plot-1, fig.width=6.5, fig.height=1.5, results="hide", eval=FALSE}
p3 <- structure_plot(fit,topics = 1:6,colors = topic_colors)
print(p3)
```

The Structure plot is effectively a stacked bar chart, in which each
bar corresponds to a cell, and each colour corresponds to a factor (or
"topic"). Since the mixture proportions for each cell sum to 1, the
stacked bars all have a height of 1.

The colours in the Structure plot were chosen to correspond to the
clusters shown in the plots above.

The clustering of the cells into B cells (blue), CD14+ monocytes
(green) and CD34+ cells (purple) is evident from this plot; they are
each largely characterized by a single factor. The NK and T cells is
less distinct, and this is reflected by more sharing of factors (in
particular, factor 6) among NK and T cells.

The versality of the Structure plot can also present a challenge,
particularly for cell populations with more complex or fine-scale
structure. It may therefore take some effort and creativity to create
a Structure plot that effectively highlights the structure in the
data. A simple way we can improve the Structure plot here is to
arrange the cells by their assignment to the 5 clusters:

```{r structure-plot-2, fig.width=6.5, fig.height=1.5, results="hide", message=FALSE, eval=FALSE}
p4 <- structure_plot(fit,topics = 1:6,colors = topic_colors,
                     grouping = pbmc_4k$samples$subpop,gap = 25)
print(p4)
```

This second Structure plot better highlights heterogeneity within the
clusters, particularly the T cells. Within the T cell cluster, there
is a distinctive subset of cells with sizeable contributions from the
"NK cells" factor (factor 1, light blue). These are the same cells
that sit in between the dense clumps of NK and T cells in the PCA plot
above. To highlight this subset of cells---which turn out to be CD8+
cells---we define a new cluster, and create a new Structure plot with
the 6 clusters:

```{r structure-plot-3, fig.width=6.5, fig.height=1.5, results="hide", message=FALSE, eval=FALSE}
clusters <- as.character(clusters)
clusters[clusters == "T" & fit_multinom$L[,1] > 0.1] <- "CD8+"
clusters <- factor(clusters)
p5 <- structure_plot(fit_multinom,topics = 1:6,colors = topic_colors,
                     grouping = clusters,gap = 25)
print(p5)
```

Even if the boundaries between the less distinct T, NK and CD8+
clusters are somewhat arbitrary, subdividing these cells into clusters
is nonetheless useful for guiding the eye toward interesting patterns
in the mixture proportions, such as the continuous variation in topics
5 and 6 within the T cells cluster.

Annotate clusters by top differentially expressed genes
-------------------------------------------------------

In this remaining parts of the analysis, we perform a differential
expression analysis, and use the results of this analysis to annotate
clusters and factors ("topics"). We begin by analyzing differential
expression in the 6 clusters.

Since `diff_count_analysis` in fastTopics performs differential
expression analysis for a multinomial topic model, here we implement
the analysis for the clusters by treating the clustering as a topic
model in which all the mixture proportions are 0 or 1.

```{r diff-count-analysis-clusters, eval=FALSE}
diff_count_clusters <- diff_count_clusters(counts,clusters)
```

The results of the differential expresison analysis can be visualized
using a "volcano plot", which shows, for each gene, the (base-2)
log-fold change in the x-axis against the support for differential
expression (*z*-score magnitude) in the y-axis. In volcano plot for
the "B" cluster, top genes include *CD79A*, an established "marker
gene" for B cells:

```{r volcano-plot-bcell-cluster, fig.height=3.5, fig.width=4, eval=FALSE}
p6 <- volcano_plot(diff_count_clusters,k = "B",label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 10)
print(p6)
```

Likewise, *CD3D*, a marker gene for T cells, is one of the top genes
in the "T" cluster volcano plot:

```{r volcano-plot-tcell-cluster, fig.height=3.5, fig.width=4, eval=FALSE}
p7 <- volcano_plot(diff_count_clusters,k = "T",label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 10)
print(p7)
```

Annotate topics by top differentially expressed genes
-----------------------------------------------------

We can also use the same differential expression analysis to identify
genes that are highly differentially expressed among the topics:

```{r diff-count-analysis-topics, eval=FALSE}
diff_count_topics <- diff_count_analysis(fit,counts)
```

For example, topic 4, which is the predominant topic in the "B"
cluster, includes the B cells marker gene *CD79A* among its top genes
in the volcano plot:

```{r volcano-plot-bcell-topic, fig.height=3.5, fig.width=4, eval=FALSE}
p8 <- volcano_plot(diff_count_topics,k = 4,label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 10)
print(p8)
```

Note that the estimated differential expression (log-fold change) for
this gene, and others, is much higher in the topic; 6.5 in the cluster
vs. >10 in the topic. Due to its greater flexibility in capturing
cells as mixtures of expression patterns, the topic model often better
isolates expression patterns specific to a cell type.

The largest of the 6 clusters captures T cells. Unlike the B cells
cluster, T cells are represented as a mixture of two topics. After
merging topics 5 and 6, we can perform differential expression
analysis on the combined topic and create a volcano plot highlighting
the top differentially expression genes:

```{r volcano-plot-tcell-topic, fig.height=3.5, fig.width=4, results="hide", eval=FALSE}
diff_count_merged <- diff_count_analysis(merge_topics(fit_multinom,5:6),counts)
p9 <- volcano_plot(diff_count_merged,k = "k5+k6",label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 10)
print(p9)
```

Like the B cells topic, the estimated log-fold change for the T-cell
marker gene *CD3D* is much greater in the topic (>10) than the "T"
cluster (3.5).

The topics also capture continuous variation *within* the T-cell
cluster:

```{r volcano-plot-cd4, fig.height=3.5, fig.width=4, results="hide", eval=FALSE}
i <- which(clusters == "T")
fit_t <- select_loadings(fit_multinom,loadings = i)
diff_count_t <- diff_count_analysis(fit_t,counts[i,])
p10 <- volcano_plot(diff_count_t,k = 5,label_above_quantile = 0.995,
                   labels = genes$symbol,max.overlaps = 16)
print(p10)
```

Judging top differentially expression genes such as *CD4* and *CCR10*,
this topic likely corresponds to CD4+/CD8+ lineage differentiation in
T cells.

An interactive volcano plot
---------------------------

When the volcano plot shows many overlapping differentially expressed
genes, like it does here for topic 5 in the T-cells cluster, it can be
helpful to explore the results interactively. The function
`volcano_plotly` creates an interactive plot that can be viewed in a
Web browser:

```{r volcano-plotly, eval=FALSE}
p11 <- volcano_plotly(diff_count_t,k = 5,"volcano_plot_t_cells.html",
                      labels = genes$symbol)
```

The interactive volcano plot can also be viewed
[here](volcano_plot_t_cells.html).

Session info
------------

This is the version of R and the packages that were used to generate
these results.

```{r session-info}
sessionInfo()
```

[zheng-2017]: https://doi.org/10.1038/ncomms14049

