---
title: "Analysis of single-cell RNA-seq data using a topic model, Part 1: basic concepts"
author: Peter Carbonetto
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: no
    highlight: textmate
    theme: readable
vignette: >
  %\VignetteIndexEntry{Analysis of single-cell RNA-seq data using a topic model, Part 1: basic concepts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The aim of this vignette is to introduce the basic concepts behind
an analysis of single-cell RNA-seq data using a topic model. 

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

We begin our analysis by loading the packages. Then we set the seed so
that the results can be reproduced.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Matrix)
library(fastTopics)
library(ggplot2)
library(cowplot)
set.seed(1)
```

The data set
------------

We will illustrate the concepts using a single-cell RNA-seq data set
from [Zheng *et al* (2017)][zheng-2017]. These data are reference
transcriptome profiles from 10 bead-enriched subpopulations of
peripheral blood mononuclear cells (PBMCs). The original data set is
much larger---for this introduction, we have taken a small subset of
approximately 3,700 cells.

The data we will analyze are the unique molecular identifier (UMI)
counts. These data are stored as an $n \times m$ sparse matrix, where
$n$ is the number of cells and $m$ is the number of genes:

```{r load-data}
data(pbmc_4k)
counts <- pbmc_4k$counts
dim(counts)
```

The UMI count data are expected to be "sparse"---that is, most of the
counts are expected to be zero. Indeed, over 95% of the UMI counts are
zero:

```{r nonzeros}
mean(counts > 0)
```

No need to normalize or transform
---------------------------------

Most analyses single-cell RNA-seq data involve a pre-processing step
in which the UMI counts are log-transformed and normalized. *When
analyzing UMI counts with a topic model, you should not normalize or
transform the counts.* In particular:

1. There is no need to log-transform the UMI counts because the counts
   are modeled directly using the multinomial distribution.

2. There is no need to normalize the counts in each cell by sequencing
   coverage because the multinomial is *conditioned* on the cell's
   total UMI count.

Additionally, any analyses select only the most highly variable
genes. *We recommend instead using all genes.* The only exception are
genes in which all the UMI counts are zero---these genes should be
removed prior to a topic model analysis.

Fit the topic model
-------------------

Since no pre-processing is needed, we can move directly to the next
step of the analysis: fitting the topic model to the UMI count
data. This is accomplished by a single call to `fit_topic_model`:

```{r fit-topic-model, eval=FALSE}
fit <- fit_topic_model(counts,k = 6)
```

This is typically the most computationally burdensome step---it may
take several minutes to run `fit_topic_model` on this data set. For
convenience, we saved the output from this call:

```{r load-fit}
fit <- pbmc_4k$fit
```

To fit a topic model, we must specify $K$, the number of topics.
Here, we have chosen $K = 6$ topics. In most settings, a good choice
of $K$ will not be known in advance, so you will you want to explore
the results from topic models at different settings of $K$.

There are several complexities to topic model fitting---we do not
elaborate on these complexities here. The `fit_topic_model` interface
is intended to hide most of these complexities, and it should work
well for a wide range of data sets. However, larger or more
challenging data sets my require some fine-tuning of the model
fitting. Model fitting is covered in more detail in a separate
vignette.

Each cell is represented as a unique mixture
--------------------------------------------

A key feature of the topic model is that each cell $i$ is represented
as a *unique mixture of the topics.* Therefore, each cell can be
summarized from the $K$ mixture proportions. This cell-specific
mixture is learned from the data. In fastTopics, the mixture
proportions for all cells are stored as an $n \times K$ matrix:

```{r loadings-1}
dim(fit$L)
```

To illustrate, here is a cell in which the pattern of expression is
almost fully captured by the fourth topic:

```{r loadings-2}
rows <- "GATATATGTCAGTG-1-b_cells"
round(fit$L[rows,],digits = 3)
```

Here are two more examples:

```{r loadings-3}
rows <- c("GACAGTACCTGTGA-1-memory_t",
          "TGAAGCACACAGCT-1-b_cells")
round(fit$L[rows,],digits = 3)
```

The third example is interesting because the observed expression is
best captured by a *mixture* of topics 4 and 6.

To make sense of these results, we first need to understand the
biological relevance of the topics. We turn to this question next.

Interpreting topics using available cell labels
-----------------------------------------------

In some cases, you may have additional information about the cells,
such as the tissue the cells were sampled from. In the PBMC data, the
cells were labeled using fluorescence-activated cell sorting
(FACS). Each cell is assigned one of five labels (each corresponding
to a cell type): B cells, CD14+ monocytes, CD34+ cells, natural killer
(NK) cells, and T cells.

```{r summary-subpop}
samples <- pbmc_4k$samples
summary(samples$subpop)
```

To visualize the relationship between the cell labels and the mixture
proportions, we create a Structure plot:

```{r structure-plot-with-celltype-labels, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
topic_colors <- c("skyblue","forestgreen","darkmagenta","dodgerblue",
                  "gold","darkorange")
p1 <- structure_plot(fit,topics = 1:6,colors = topic_colors,
                     grouping = samples$subpop,gap = 25)
print(p1)
```

The Structure plot is a simply stacked bar chart, in which each topic
is represented as a bar of a different colour. Being *proportions*,
the mixture proportions for each cell must sum to 1, so the total
height of the bars is the same for all cells, which makes it easier to
compare across cells. Patterns begin to emerge when the cells are
arranged so that cells with similar mixture proportions are positioned
close to each other. This arrangement is automated in `structure_plot`.

From this Structure plot, it is evident that topics 1 (lighter blue), 2
(green) and 4 (darker blue) closely correspond to the NK, CD14+ and B
cell types, respectively; the expression in these cells is largely
explained by a single topic.  NK cells show more heterogeneity in
expression than the CD14+ and B cells.

The mixture proportions uncover some surprises in the cells labeled as
"CD34+": while their expression is largely explained by topic 3
(purple), some cells appear to be mislabeled---they should have been
labeled as B cells or CD14+ cells. Other cells are best fit by a
mixture of several topics.

Finally, the topic model captures interesting substructure within the
T cells. Two patterns stand out from the Structure plot: first, most T
cells are a mixture of two topics, 5 (yellow) and 6 (orange), with
wide variation in the mixing proportions for these two topics; second,
there is a distinctive subset of T cells that is represented as a
mixture of three topics (topics 1, 5 and 6). To help us understand
what aspects of T cell diversity are captured by these topics, we need
to first understand how topics capture shared patterns of gene
expression.

Topics capture patterns of relative expression
----------------------------------------------

Above, we explained that the topic model represents cells as mixtures
of topics. In our running example, some of the topics seem to have a
straightforward interpretation as known cell types. More precisely,
each topic captures an expression pattern, encoded as a vector of $m$
relative expression levels ($m$ is the number of genes). In
fastTopics, the relative expression levels for all topics are stored
as an $m \times K$ matrix:

```{r factors-1}
dim(fit$F)
```

Because the expression patterns are encoded by *relative expression
levels*, the estimates of relative expression are directly comparable
across topics. For example, gene *CD79B* is an important component of
B cell differentiation, so we would expect higher expression in B
cells. Indeed, this is what we find in the topic 4 (we multiplied the
estimates by $10^6$ only to make them easier to read):

```{r factors-2}
genes <- pbmc_4k$genes
j <- which(genes$symbol == "CD79B")
round(1e6 * fit$F[j,])
```

Expression of *CD79A*---a known marker gene for B cells---provides an
even stronger clue, as it is uniquely expressed in topic 4:

```{r factors-3}
j <- which(genes$symbol == "CD79A")
round(1e6 * fit$F[j,])
```

By contrast, genes that are highly expressed in all topics, e.g.,
*COX4I1*, provide little useful information:

```{r factors-4}
j <- which(genes$symbol == "COX4I1")
round(1e6 * fit$F[j,])
```

The larger point here is that the most informative genes are the genes
with higher expression in one topic compared to the other topics (and
more generally, genes with large differences in gene expression).
This idea of analyzing differences in gene expression analysis is an
old idea. The next section explains how differential expression
analysis is performed using the topic model.

Interpreting topics by analyzing gene expression differences
------------------------------------------------------------

To set the stage for differential expression analysis using a topic
model, we begin with a "classic" differential expression analysis
using the provided cell type labels. This analysis centers on
calculation of the log-fold change statistic. For a given group or
cluster, it is defined as the (base-2) log-ratio of the average
expression (meam UMI count) among cells belonging to the group over
the average expression among cells not belonging to the group. For
example, genes *CD79A* and *CD79B* are overexpressed in B cells, by a
factor of $2^5 = 32$ times, or more; by contrast these two genes are
underexpressed in all other cell types:

```{r diff-count-analysis-1}
out1 <- diff_count_clusters(counts,samples$subpop)
rbind(out1$beta[genes$symbol == "CD79A",],
      out1$beta[genes$symbol == "CD79B",])
```

We have extended the classic differential expression analysis in a
natural way to allow for *partial membership* to groups. This
generalized differential expression analysis is implemented in the
`diff_count_analysis` function:

```{r diff-count-analysis-2}
out2 <- diff_count_analysis(fit,counts)
```

As in the standard analysis, we obtain log-fold change estimates:

```{r diff-count-analysis-3}
rbind(out2$beta[genes$symbol == "CD79A",],
      out2$beta[genes$symbol == "CD79B",])
```

These log-fold change statistics can be interpreted in the same way
(and in fact the standard log-fold change statistics emerge as a
special case of the topic-model-based differential expression analysis
when the all mixture proportions are 0 or 1). The log-fold change
estimates are noticeably larger in the "B cells" topic, and
particularly so for *CD79A*. In fact, *CD79A* is uniquely expressed in
topic 4, so the ratio essentially has a zero denominator; since we are
restricted to finite numeric calculations, a log-fold change statistic
of 10 practically means "10 or larger" (by default). Due to its
greater flexibility in capturing cells as mixtures of expression
patterns, the topic model is often more adept at isolating expression
patterns specific to a cell type.

Up to this point, we have illustrated the differential expression
analysis with known cell-type-specific genes such as *CD79A*, but
ultimately we would like to discover genes the relevant to a given
topic of known or unknown biological relevance.

The natural thing to do would be to rank genes by log-fold change. The
complication is that very lowly expressed genes could also produce, by
chance, large log-fold changes, and we would not want to rank these
genes highly. The *volcano plot* provides a visual ranking balancing
these two concerns: for each gene, the log-fold change is shown on the
x-axis, and the *p*-value or *z*-score giving a measure of support for
differential expression is shown on the *y*-axis:

```{r volcano-plot-b, fig.height=3.5, fig.width=4.5}
p2 <- volcano_plot(out2,k = 4,label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 15)
print(p2)
```

Typically the most interesting genes are found in the top-right
portion of the volcano plot---that is, genes with large log-fold
change and strong support. Indeed, many known B-cell genes such as
*CD79A*, *CD79B* and *CD19* approach the top-right corner of the
volcano plot for topic 4.

Likewise, natural killer genes such as *NKG7* emerge at the top of the
volcano plot for topic 1:

```{r volcano-plot-nk, fig.height=3.5, fig.width=4.5}
p3 <- volcano_plot(out2,k = 1,label_above_quantile = 0.998,
                   labels = genes$symbol,max.overlaps = 15)
print(p3)
```

Interpreting the topics that capture variation *within* T cells is
slightly more involved, so we old off on this investigation until
Part 2.

Evaluating the topic model
--------------------------

Explain that we can evaluate the fit via the likelihood.

```{r loglik-1}
loglik_topics <- loglik_multinom_topic_model(counts,fit)
```

*Add text here.*

```{r loglik-2}
fit_clusters <- init_poisson_nmf_from_clustering(counts,samples$subpop)
fit_clusters <- poisson2multinom(fit_clusters)
loglik_clusters <- loglik_multinom_topic_model(counts,fit_clusters)
```

*Add text here.*

```{r loglik-3, fig.width=4, fig.height=3}
i <- which(samples$subpop == "CD34+")
p4 <- qplot(x = loglik_clusters[i],y = loglik_topics[i],color = fit$L[i,3]) +
  geom_abline(slope = 1,intercept = 0,linetype = "dotted") +
  scale_color_gradient2(high = "deepskyblue",mid = "gold",low = "orangered",
                        midpoint = 0.5) +
  labs(x = "clustering",y = "topic model",color = "topic 3") +
  theme_cowplot(font_size = 12)
print(p4)
```

*Add text here.*

Session info
------------

This is the version of R and the packages that were used to generate
these results.

```{r session-info}
sessionInfo()
```

[zheng-2017]: https://doi.org/10.1038/ncomms14049

