\documentclass[final]{siamart171218}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{bm}

\setlength{\oddsidemargin}{0.65in}
\setlength{\evensidemargin}{0.65in}

\title{Topic modeling analysis of differential gene expression}

\author{Peter Carbonetto\thanks{Dept. of Human Genetics and the Research Computing Center, University of Chicago, Chicago, IL}}

\begin{document}

\maketitle

\section{Differential gene expression}

The ``log-fold change'' statistic is commonly used in microarray and
RNA sequencing experiments to quantify expression differences between
two conditions (e.g., \cite{cui-churchill-2003, quackenbush-2002}). To
motivate the ideas below, I write the log-fold change for gene $j$ and
condition $k$ as a ratio of two conditional expectations,
\begin{equation}
\mathsf{lfc}(j,k) \equiv
\log_2 \frac{E[\,x_j \,|\, \mathrm{condition} = k\,]}
            {E[\,x_j \,|\, \mathrm{condition} \neq k\,]},
\end{equation}
where $x_j$ is the measured expression level (e.g., UMI count) of gene
$j$. In experiments where the conditions are inferred---for example,
by running a machine learning algorithm to cluster the expression
profiles---this quantity could represent the difference in gene
expression between cells inside and outside a
cluster.\footnote{Supposing $n_k$ out of a total of $n$ gene
  expression profiles (cells) are from condition $k$, then
  $\mathsf{lfc}(j,k)$ can be computed as $\mathsf{lfc}(j,k) = \log_2
  \{n_{jk}/(n_j - n_{jk}) \times (n - n_k)/n_k \}$, where $n_j$ is the
  total expression of gene $j$ among all expression profiles, and
  $n_{jk}$ is the total expression of $j$ among all cells in condition
  (or cluster) $k$.}

The aim of the next sections is to define a analogue to the log-fold
change statistic for topic modeling.

\section{The multinomial topic model and Poisson non-negative matrix
  factorization}

Here we briefly describe the multinomial topic model, and its
connection to Poisson non-negative matrix factorization (Poisson NMF).

We begin with the ``bag of words'' description, which was used to
describe the LDA model \cite{blei-2003}. In this view, each document
(or gene expression profile) $i$ is represented as a vector of
terms/genes, $w_i = (w_{i1}, \ldots, w_{is_i})$, where $s_i$ is the
size of document $i$. (The order of the words or genes appearing in
this vector doesn't matter, hence the ``bag of words.'') Each $w_{it}
\in \{1, \ldots, m\}$ is term/gene $j$ with probability
$p(w_{it} = j \,|\, z_{it} = k) = f_{jk}$, in which we have introduced
$z_{it}$, a variable indicating which topic $k \in \{1, \ldots, K\}$
the word/gene is drawn from. The topic indicator variables for
document $i$ are in turn generated according to $p(z_{it} = k) =
l_{ik}$.

This process also defines a {\em multinomial} model for an $n \times
m$ matrix of counts $x_{ij}$:
\begin{equation}
x_{i1}, \ldots, x_{im} \sim
\mathrm{Multinom}(x_{i1}, \ldots, x_{im}; s_i, \pi_i),
\end{equation}
where $x_{ij} = \sum_{t=1}^{s_i} \delta_j(w_{it})$ is the number of
times term/gene $j$ appears in document/cell $i$, and the
probabilities $\pi_{ij}$ are weighted sums of the ``factors''
$f_{jk}$,
\begin{equation}
\pi_{ij} = \sum_{k=1}^K l_{ik} f_{jk}.
\end{equation}
The log-likelihood for the multinomial topic model, ignoring terms
that do not depend on the model parameters, has a simple expression:
\begin{equation}
\log p(x) = \sum_{i=1}^n \sum_{j=1}^m
x_{ij} \log({\textstyle \sum_{k=1}^K l_{ik} f_{jk}}).
\end{equation}

As we have shown elsewhere, the multinomial topic model is closely
related to a Poisson non-negative matrix factorization of the count
data,
\begin{equation}
x_{ij} \sim \mathrm{Poisson}(\lambda_{ij}),
\end{equation}
where $\lambda_{ij} = \sum_{k=1}^K \hat{l}_{ik} \hat{f}_{jk}$. Given a
Poisson NMF fit, an equivalent multinomial topic model can be easily
recovered, as we have shown elsewhere.

\section{Gene expression differences in topics}

Returning to the question of assessing differential gene expression,
there are two new twists when done in the context of topic modeling:
\begin{enumerate}
  
\item The cluster (topic) assignments are probabilistic.

\item The cluster assignments are made at the level of genes, not
  cells.

\end{enumerate}
I propose a log-fold change statistic to address these two points. It
compares the probability of gene $j$ occurring ($w = j$) given topic
$k$ ($z = k$) versus the probability given assignment a topic other
than $k$ ($z \neq k$):
\begin{equation}
\mathsf{lfc}(j,k) \equiv
\log_2 \frac{p(w = j \,|\, z = k)}
            {p(w = j \,|\, z \neq k)}.
\end{equation}
For a given gene $j$ and topic $k$, $\mathsf{lfc}(j,k)$ can be
calculated as
\begin{align}
\mathsf{lfc}(j,k) &=
\log_2 \left\{ \frac{p(w = j, z = k)}
                    {p(w = j, z \neq k)} \times
               \frac{p(z \neq k)}{p(z = k)} \right\} \nonumber \\
&= \log_2 \left\{ 
\frac{\sum_{i=1}^n \sum_{t=1}^{s_i} \delta_j(w_{it}) \, \phi_{ijkt}}
     {\sum_{i=1}^n \sum_{t=1}^{s_i} \delta_j(w_{it}) (1 - \phi_{ijkt})}
\times \frac{\sum_{i=1}^n \sum_{j'=1}^m \sum_{t=1}^{s_i} 
             \delta_{j'}(w_{it}) (1-\phi_{ij'kt})}
            {\sum_{i=1}^n \sum_{j'=1}^m \sum_{t=1}^{s_i} \delta_{j'}(w_{it}) 
             \, \phi_{ij'kt}} \right\},
\label{eq:lfc-topic-model}
\end{align}
where $\phi_{ijkt}$ denotes the posterior probability of $z_{it} = k$
given $w_{it} = j$,
\begin{align}
\phi_{ijkt} &\equiv p(z_{it} = k \,|\, w_{it} = j) \nonumber \\
&= \frac{p(w_{it} = j \,|\, z_{it} = k) \, p(z_{it} = k)}
        {\sum_{k'=1}^K p(w_{it} = j \,|\, z_{it} = k') \, p(z_{it} = k')} 
   \nonumber \\
&= \frac{l_{ik} f_{jk}}
        {\sum_{k'=1}^K l_{ik'} f_{jk'}}.
\end{align}
Since the topic assignments $z_{it}$ do not depend on $t$---that is,
we can drop the ``$t$'' subscript from the $\phi_{ijkt}$'s---the expression
for the {\em lfc} simplifies:
\begin{equation}
\mathsf{lfc}(j,k) = \log_2 \left\{ 
\frac{\sum_{i=1}^n x_{ij} \, \phi_{ijk}}
     {\sum_{i=1}^n x_{ij} (1 - \phi_{ijk})}
     \times \frac{\sum_{i=1}^n \sum_{j'=1}^m x_{ij'} (1-\phi_{ij'k})}
                 {\sum_{i=1}^n \sum_{j'=1}^m x_{ij'} \phi_{ij'k}} \right\}.
\end{equation}

At the maximum-likelihood solution (MLE) of the $l_{ik}$'s and $f_{kl}$'s,
the {\em lfc} statistic simplifies further:
\begin{equation}
\mathsf{lfc}(j,k) = 
\log_2 \left\{ \frac{\sum_{i=1}^n x_{ij} \, \phi_{ijk}}
                    {\sum_{i=1}^n x_{ij} (1 - \phi_{ijk})} \times
               \frac{\sum_{i=1}^n s_i (1 - l_{ik})}
                    {\sum_{i=1}^n s_i l_{ik}} \right\}.
\end{equation}
This is because, at the MLE, the loadings $s_{ik}$, $k = 1, \ldots,
K$, for a given document/cell $i$ should be proportional to the sums
$\sum_{j=1}^m x_{ij} \phi_{ijk}$.

Finally, it is convenient that the {\em lfc}
\eqref{eq:lfc-topic-model} will be the same if we replace the
multinomial topic model parameters $l_{ik}$ and $f_{jk}$ with the
corresponding parameters of the Poisson NMF, $\hat{l}_{ik}$ and
$\hat{f}_{jk}$ (proof not given).

\bibliographystyle{siamplain}
\bibliography{lfoldchange}

\end{document}

