<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of single-cell RNA-seq data using fastTopics: an illustration • fastTopics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of single-cell RNA-seq data using fastTopics: an illustration">
<meta property="og:description" content="fastTopics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3-191</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/fastTopics">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of single-cell RNA-seq data using fastTopics: an illustration</h1>
                        <h4 class="author">Peter Carbonetto</h4>
            
            <h4 class="date">2020-12-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/fastTopics/blob/master/vignettes/single_cell_rnaseq_demo.Rmd"><code>vignettes/single_cell_rnaseq_demo.Rmd</code></a></small>
      <div class="hidden name"><code>single_cell_rnaseq_demo.Rmd</code></div>

    </div>

    
    
<p>Here we illustrate the use of fastTopics to infer structure from single-cell RNA-seq data. A basic analysis consists of the following steps:</p>
<ol style="list-style-type: decimal">
<li><p>Load and prepare the count data.</p></li>
<li><p>Fit a Poisson non-negative matrix factorization (NMF) to the UMI count data.</p></li>
<li><p>Recover the multinomial topic model—also called the multinomial “grade-of-membership” model—from the Poisson NMF model fit.</p></li>
<li><p>Analyze structure in the estimated mixture proportions matrix.</p></li>
<li><p>Annotate selected clusters and factors by examining genes that are highly differentially expressed among clusters or factors.</p></li>
</ol>
<p>In this vignette, we walk through each of these steps and demonstrate how to implement these steps using fastTopics. In other exploratory analyses, these steps might not be strictly sequential—for example, the annotation in Step 5 may inform the exploration of substructure in Step 4.</p>
<p>Begin by loading the packages used below. The ggplot2 and cowplot packages are used to create some of the plots.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Matrix</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">fastTopics</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">cowplot</span>)</pre></body></html></div>
<p>Set the seed so that the results can be reproduced.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)</pre></body></html></div>
<div id="load-and-prepare-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-and-prepare-data" class="anchor"></a>Load and prepare data</h2>
<p>Load the single-cell RNA-seq data:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">pbmc_4k</span>)
<span class="no">samples</span> <span class="kw">&lt;-</span> <span class="no">pbmc_4k</span>$<span class="no">samples</span>
<span class="no">genes</span>   <span class="kw">&lt;-</span> <span class="no">pbmc_4k</span>$<span class="no">genes</span>
<span class="no">counts</span>  <span class="kw">&lt;-</span> <span class="no">pbmc_4k</span>$<span class="no">counts</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">counts</span>)
<span class="co"># [1]  3774 21952</span></pre></body></html></div>
<p>This data set is made up of reference transcriptome profiles that were generated from 10 different bead-enriched subpopulations of PBMCs (Donor A), as described in <a href="https://doi.org/10.1038/ncomms14049">Zheng <em>et al</em> (2017)</a>. The UMI counts are stored as an <span class="math inline">\(n \times m\)</span> sparse matrix, where <span class="math inline">\(n\)</span> is the number of cells and <span class="math inline">\(m\)</span> is the number of genes.</p>
<p>No normalization or transformation of the UMI counts is needed. The only pre-processing step we take is to remove genes that are not expressed in any of the cells.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">j</span>      <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">counts</span> <span class="kw">&gt;</span> <span class="fl">0</span>) <span class="kw">&gt;=</span> <span class="fl">1</span>)
<span class="no">genes</span>  <span class="kw">&lt;-</span> <span class="no">genes</span>[<span class="no">j</span>,]
<span class="no">counts</span> <span class="kw">&lt;-</span> <span class="no">counts</span>[,<span class="no">j</span>]
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">counts</span>)
<span class="co"># [1]  3774 16791</span></pre></body></html></div>
<p>As expected, most (&gt;95%) of the UMI counts are zero:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">counts</span> <span class="kw">&gt;</span> <span class="fl">0</span>)
<span class="co"># [1] 0.04265257</span></pre></body></html></div>
</div>
<div id="fit-poisson-nmf-model" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-poisson-nmf-model" class="anchor"></a>Fit Poisson NMF model</h2>
<p>By exploiting the relationship between the multinomial topic model and non-negative matrix factorization (we have a vignette elaborating on this relationship), we leverage fast NMF optimization algorithms to fit the multinomial topic model.</p>
<p>We chose <span class="math inline">\(K = 6\)</span> factors. In practice, you will likely want to explore the results from different settings of <span class="math inline">\(K\)</span>.</p>
<p>Since the model fitting can take several minutes to run, or perhaps more, we have saved a pre-fitted model:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">fit</span> <span class="kw">&lt;-</span> <span class="no">pbmc_4k</span>$<span class="no">fit</span></pre></body></html></div>
<p>If you would like to see the model fitting in action, run this code:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">fit0</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fit_poisson_nmf.html">fit_poisson_nmf</a></span>(<span class="no">counts</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="fl">6</span>,<span class="kw">numiter</span> <span class="kw">=</span> <span class="fl">100</span>,<span class="kw">method</span> <span class="kw">=</span> <span class="st">"em"</span>,
                        <span class="kw">control</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">numiter</span> <span class="kw">=</span> <span class="fl">4</span>,<span class="kw">nc</span> <span class="kw">=</span> <span class="fl">2</span>))
<span class="no">fit</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fit_poisson_nmf.html">fit_poisson_nmf</a></span>(<span class="no">counts</span>,<span class="kw">fit0</span> <span class="kw">=</span> <span class="no">fit0</span>,<span class="kw">numiter</span> <span class="kw">=</span> <span class="fl">100</span>,<span class="kw">method</span> <span class="kw">=</span> <span class="st">"scd"</span>,
                        <span class="kw">control</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">extrapolate</span> <span class="kw">=</span> <span class="fl">TRUE</span>,<span class="kw">numiter</span> <span class="kw">=</span> <span class="fl">4</span>,<span class="kw">nc</span> <span class="kw">=</span> <span class="fl">2</span>))</pre></body></html></div>
<p>(This is the code that was used to generate the stored result.) Before running this code, may want to change <code>control$nc</code>. This controls the number of threads in the TBB multithreaded computations, and for best performance this should be set to the number of availablle processors.</p>
<p>Judging by this plot which shows the change in the log-likelihood over time, the iterates get close to a maxium-likelihood solution after about 150 iterations.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_progress_poisson_nmf.html">plot_progress_poisson_nmf</a></span>(<span class="no">fit</span>,<span class="kw">x</span> <span class="kw">=</span> <span class="st">"iter"</span>,<span class="kw">add.point.every</span> <span class="kw">=</span> <span class="fl">10</span>,
                          <span class="kw">colors</span> <span class="kw">=</span> <span class="st">"black"</span>) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span>(<span class="kw">font_size</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/plot-loglik-1.png" width="420" style="display: block; margin: auto;"></p>
<p><strong>Some general guidance on model fitting.</strong> We have found that the EM algorithm initially makes rapid progress toward a solution, but this convergence slows considerably as the iterates approach the solution. Close to a solution, we have found that other algorithms make much more rapid progress than EM. Therefore, we have taken a two-stage approach to model fitting: first, we perform EM updates; second, we refine the fit using extrapolated SCD updates (we also tried extrapolated CCD updates, and updates without extrapolation, and we found that the extrapolated SCD updates almost always worked best). For larger data sets, more than 200 updates may be needed to obtain a good fit.</p>
</div>
<div id="recover-multinomial-topic-model" class="section level2">
<h2 class="hasAnchor">
<a href="#recover-multinomial-topic-model" class="anchor"></a>Recover multinomial topic model</h2>
<p>Once we have fit a Poisson NMF model, the multinomial topic model is easly recovered by a reparameterization:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">fit_multinom</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/poisson2multinom.html">poisson2multinom</a></span>(<span class="no">fit</span>)</pre></body></html></div>
</div>
<div id="identify-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h2>
<p>The topic or grade-of-membership model is able to capture a variety of structure, from discrete (clusters) to continuous. We begin by showing how the model can be used to identify clusters.</p>
<p>Here we apply the simplest clustering method, <em>k</em>-means, to the mixture proportions matrix, <span class="math inline">\(L\)</span>. We experimented with different numbers of clusters and we found that 6 clusters worked best in this example. Note that <em>k</em>-means is sensitive to (random) initialization, so we set the seed.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(<span class="no">fit_multinom</span>$<span class="no">L</span>,<span class="kw">centers</span> <span class="kw">=</span> <span class="fl">6</span>,<span class="kw">iter.max</span> <span class="kw">=</span> <span class="fl">100</span>)$<span class="no">cluster</span></pre></body></html></div>
<p>Note that, in some cases, it may be better to apply <em>k</em>-means to the principal components (PCs) of the mixture proportions, rather than the mixture proportions directly. The PCs can easily be computed using <code>prcomp</code>.</p>
<p>You will find that some visual inspection followed by manual refinement may be needed to get the “right” clustering, regardless of whether <em>k</em>-means or another clustering algorithm is used. By visual inspection, we found that <em>k</em>-means split the largest cluster into two subclusters. After merging these two subclusters, we ended up with 5 clusters:</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">clusters</span>[<span class="no">clusters</span> <span class="kw">==</span> <span class="fl">3</span>] <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">clusters</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">clusters</span>)
<span class="co">#    1    2    4    5    6 </span>
<span class="co"># 1447  207  797  616  707</span></pre></body></html></div>
<p>With the benefit of hindsight, we know that these 5 clusters correspond to 5 cell types: B cells, T cells, natural killer (NK) cellls, CD14+ monoytes and CD34+ cells. (Below we show how to annotate these clusters by performing a differential expression analysis.) For clarity of presentation, we label the clusters by these cell types, noting that, in most practical settings, labeling the clusters is not usually not possible until downstream analyses have been conducted.</p>
<p>The clustering of the cells is mostly evident from the top two PCs of the mixture proportions:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="no">clusters</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"T"</span>,<span class="st">"CD14+"</span>,<span class="st">"B"</span>,<span class="st">"CD34+"</span>,<span class="st">"NK"</span>)
<span class="no">cluster_colors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gold"</span>,<span class="st">"forestgreen"</span>,<span class="st">"dodgerblue"</span>,<span class="st">"darkmagenta"</span>,<span class="st">"skyblue"</span>)
<span class="no">p1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/pca_plot.html">pca_plot</a></span>(<span class="no">fit_multinom</span>,<span class="kw">fill</span> <span class="kw">=</span> <span class="no">clusters</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="no">cluster_colors</span>)
<span class="no">p2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/pca_plot.html">pca_hexbin_plot</a></span>(<span class="no">fit_multinom</span>,<span class="kw">bins</span> <span class="kw">=</span> <span class="fl">24</span>)
<span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="no">p1</span>,<span class="no">p2</span>,<span class="kw">rel_widths</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">9</span>,<span class="fl">10</span>))</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/pca-1-1.png" width="750" style="display: block; margin: auto;"></p>
<p>Examining the projection of the cells onto PCs 1 and 2, the B cells cluster is the most distinct, while the NK and T cell clusters are much less distinct, and arguably it is incorrect to call them “clusters”. (The CD14+ and CD34+ cells separate much more clearly in PCs 3 and 4.)</p>
<p>The PCA “hexbin” density plot is most helpful when there are many overlapping data points in the PCA plot.</p>
</div>
<div id="visualize-mixture-proportions-using-a-structure-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-mixture-proportions-using-a-structure-plot" class="anchor"></a>Visualize mixture proportions using a Structure plot</h2>
<p>A much more versatile tool for exploring the mixture proportions is the Structure plot. Here is a basic example:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">topic_colors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"skyblue"</span>,<span class="st">"forestgreen"</span>,<span class="st">"darkmagenta"</span>,<span class="st">"dodgerblue"</span>,
                  <span class="st">"gold"</span>,<span class="st">"darkorange"</span>)
<span class="no">p3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span>(<span class="no">fit_multinom</span>,<span class="kw">topics</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">6</span>,<span class="kw">colors</span> <span class="kw">=</span> <span class="no">topic_colors</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p3</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/structure-plot-1-1.png" width="780" style="display: block; margin: auto;"></p>
<p>The Structure plot is effectively a stacked bar chart, in which each bar corresponds to a cell, and each colour corresponds to a factor (or “topic”). Since the mixture proportions for each cell sum to 1, the stacked bars all have a height of 1.</p>
<p>The colours in the Structure plot were chosen to correspond to the clusters shown in the plots above.</p>
<p>The clustering of the cells into B cells (blue), CD14+ monocytes (green) and CD34+ cells (purple) is evident from this plot; they are each largely characterized by a single factor. The NK and T cells is less distinct, and this is reflected by more sharing of factors (in particular, factor 6) among NK and T cells.</p>
<p>The versality of the Structure plot can also present a challenge, particularly for cell populations with more complex or fine-scale structure. It may therefore take some effort and creativity to create a Structure plot that effectively highlights the structure in the data. A simple way we can improve the Structure plot here is to arrange the cells by their assignment to the 5 clusters:</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">p4</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span>(<span class="no">fit_multinom</span>,<span class="kw">topics</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">6</span>,<span class="kw">colors</span> <span class="kw">=</span> <span class="no">topic_colors</span>,
                     <span class="kw">grouping</span> <span class="kw">=</span> <span class="no">clusters</span>,<span class="kw">gap</span> <span class="kw">=</span> <span class="fl">25</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p4</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/structure-plot-2-1.png" width="780" style="display: block; margin: auto;"></p>
<p>This second Structure plot better highlights heterogeneity within the clusters, particularly the T cells. Within the T cell cluster, there is a distinctive subset of cells with sizeable contributions from the “NK cells” factor (factor 1, light blue). These are the same cells that sit in between the dense clumps of NK and T cells in the PCA plot above. To highlight this subset of cells—which turn out to be CD8+ cells—we define a new cluster, and create a new Structure plot with the 6 clusters:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">clusters</span>)
<span class="no">clusters</span>[<span class="no">clusters</span> <span class="kw">==</span> <span class="st">"T"</span> <span class="kw">&amp;</span> <span class="no">fit_multinom</span>$<span class="no">L</span>[,<span class="fl">1</span>] <span class="kw">&gt;</span> <span class="fl">0.1</span>] <span class="kw">&lt;-</span> <span class="st">"CD8+"</span>
<span class="no">clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">clusters</span>)
<span class="no">p5</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span>(<span class="no">fit_multinom</span>,<span class="kw">topics</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">6</span>,<span class="kw">colors</span> <span class="kw">=</span> <span class="no">topic_colors</span>,
                     <span class="kw">grouping</span> <span class="kw">=</span> <span class="no">clusters</span>,<span class="kw">gap</span> <span class="kw">=</span> <span class="fl">25</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p5</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/structure-plot-3-1.png" width="780" style="display: block; margin: auto;"></p>
<p>Even if the boundaries between the less distinct T, NK and CD8+ clusters are somewhat arbitrary, subdividing these cells into clusters is nonetheless useful for guiding the eye toward interesting patterns in the mixture proportions, such as the continuous variation in topics 5 and 6 within the T cells cluster.</p>
</div>
<div id="annotate-clusters-by-top-differentially-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#annotate-clusters-by-top-differentially-expressed-genes" class="anchor"></a>Annotate clusters by top differentially expressed genes</h2>
<p>In this remaining parts of the analysis, we perform a differential expression analysis, and use the results of this analysis to annotate clusters and factors (“topics”). We begin by analyzing differential expression in the 6 clusters.</p>
<p>Since <code>diff_count_analysis</code> in fastTopics performs differential expression analysis for a multinomial topic model, here we implement the analysis for the clusters by treating the clustering as a topic model in which all the mixture proportions are 0 or 1.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">fit_clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fit_poisson_nmf.html">init_poisson_nmf_from_clustering</a></span>(<span class="no">counts</span>,<span class="no">clusters</span>)
<span class="no">diff_count_clusters</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/diff_count_analysis.html">diff_count_analysis</a></span>(<span class="no">fit_clusters</span>,<span class="no">counts</span>)
<span class="co"># All topic proportions are either zero or one; using simpler single-topic calculations for model parameter estimates</span>
<span class="co"># Fitting 16791 x 6 = 100746 univariate Poisson models.</span>
<span class="co"># Computing log-fold change statistics.</span></pre></body></html></div>
<p>The results of the differential expresison analysis can be visualized using a “volcano plot”, which shows, for each gene, the (base-2) log-fold change in the x-axis against the support for differential expression (<em>z</em>-score magnitude) in the y-axis. In volcano plot for the “B” cluster, top genes include <em>CD79A</em>, an established “marker gene” for B cells:</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">p6</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plot</a></span>(<span class="no">diff_count_clusters</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="st">"B"</span>,<span class="kw">label_above_quantile</span> <span class="kw">=</span> <span class="fl">0.998</span>,
                   <span class="kw">labels</span> <span class="kw">=</span> <span class="no">genes</span>$<span class="no">symbol</span>,<span class="kw">max.overlaps</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p6</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/volcano-plot-bcell-cluster-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Likewise, <em>CD3D</em>, a marker gene for T cells, is one of the top genes in the “T” cluster volcano plot:</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">p7</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plot</a></span>(<span class="no">diff_count_clusters</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="st">"T"</span>,<span class="kw">label_above_quantile</span> <span class="kw">=</span> <span class="fl">0.998</span>,
                   <span class="kw">labels</span> <span class="kw">=</span> <span class="no">genes</span>$<span class="no">symbol</span>,<span class="kw">max.overlaps</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p7</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/volcano-plot-tcell-cluster-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div id="annotate-topics-by-top-differentially-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#annotate-topics-by-top-differentially-expressed-genes" class="anchor"></a>Annotate topics by top differentially expressed genes</h2>
<p>We can also use the same differential expression analysis to identify genes that are highly differentially expressed among the topics:</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">diff_count_topics</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/diff_count_analysis.html">diff_count_analysis</a></span>(<span class="no">fit</span>,<span class="no">counts</span>)
<span class="co"># Fitting 16791 x 6 = 100746 univariate Poisson models.</span>
<span class="co"># Computing log-fold change statistics.</span></pre></body></html></div>
<p>For example, topic 4, which is the predominant topic in the “B” cluster, includes the B cells marker gene <em>CD79A</em> among its top genes in the volcano plot:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">p8</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plot</a></span>(<span class="no">diff_count_topics</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="fl">4</span>,<span class="kw">label_above_quantile</span> <span class="kw">=</span> <span class="fl">0.998</span>,
                   <span class="kw">labels</span> <span class="kw">=</span> <span class="no">genes</span>$<span class="no">symbol</span>,<span class="kw">max.overlaps</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p8</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/volcano-plot-bcell-topic-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Note that the estimated differential expression (log-fold change) for this gene, and others, is much higher in the topic; 6.5 in the cluster vs. &gt;10 in the topic. Due to its greater flexibility in capturing cells as mixtures of expression patterns, the topic model often better isolates expression patterns specific to a cell type.</p>
<p>The largest of the 6 clusters captures T cells. Unlike the B cells cluster, T cells are represented as a mixture of two topics. After merging topics 5 and 6, we can perform differential expression analysis on the combined topic and create a volcano plot highlighting the top differentially expression genes:</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">diff_count_merged</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/diff_count_analysis.html">diff_count_analysis</a></span>(<span class="fu"><a href="../reference/merge_topics.html">merge_topics</a></span>(<span class="no">fit_multinom</span>,<span class="fl">5</span>:<span class="fl">6</span>),<span class="no">counts</span>)
<span class="no">p9</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plot</a></span>(<span class="no">diff_count_merged</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="st">"k5+k6"</span>,<span class="kw">label_above_quantile</span> <span class="kw">=</span> <span class="fl">0.998</span>,
                   <span class="kw">labels</span> <span class="kw">=</span> <span class="no">genes</span>$<span class="no">symbol</span>,<span class="kw">max.overlaps</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p9</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/volcano-plot-tcell-topic-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Like the B cells topic, the estimated log-fold change for the T-cell marker gene <em>CD3D</em> is much greater in the topic (&gt;10) than the “T” cluster (3.5).</p>
<p>The topics also capture continuous variation <em>within</em> the T-cell cluster:</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">i</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">clusters</span> <span class="kw">==</span> <span class="st">"T"</span>)
<span class="no">fit_t</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/select_loadings.html">select_loadings</a></span>(<span class="no">fit_multinom</span>,<span class="kw">loadings</span> <span class="kw">=</span> <span class="no">i</span>)
<span class="no">diff_count_t</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/diff_count_analysis.html">diff_count_analysis</a></span>(<span class="no">fit_t</span>,<span class="no">counts</span>[<span class="no">i</span>,])
<span class="no">p10</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plot</a></span>(<span class="no">diff_count_t</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="fl">5</span>,<span class="kw">label_above_quantile</span> <span class="kw">=</span> <span class="fl">0.995</span>,
                   <span class="kw">labels</span> <span class="kw">=</span> <span class="no">genes</span>$<span class="no">symbol</span>,<span class="kw">max.overlaps</span> <span class="kw">=</span> <span class="fl">16</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p10</span>)</pre></body></html></div>
<p><img src="single_cell_rnaseq_demo_files/figure-html/volcano-plot-cd4-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Judging top differentially expression genes such as <em>CD4</em>, this topic likely corresponds to CD4+/CD8+ lineage differentiation in T cells.</p>
</div>
<div id="an-interactive-volcano-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#an-interactive-volcano-plot" class="anchor"></a>An interactive volcano plot</h2>
<p>When the volcano plot shows many overlapping differentially expressed genes, like it does here for topic 5 in the T-cells cluster, it can be helpful to explore the results interactively. The function <code>volcano_plotly</code> creates an interactive plot that can be viewed in a Web browser:</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">p11</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plotly</a></span>(<span class="no">diff_count_t</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="fl">5</span>,<span class="st">"volcano_plot_t_cells.html"</span>,
                      <span class="kw">labels</span> <span class="kw">=</span> <span class="no">genes</span>$<span class="no">symbol</span>)</pre></body></html></div>
<p>The interactive volcano plot can also be viewed <a href="volcano_plot_t_cells.html">here</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Joyce Hsiao, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
